### PR TITLE
feat: add compression metadata flatbuffer schema and tests

### PR DESCRIPTION
**Pull request by rkuester**
Add a flatbuffer schema for describing compressed models.
Flatbuffers with this schema are to be used as the value in a
.tflite model flatbuffer metadata field, and contain the extra
information necessary to describe a compressed model.

Include tests to ensure basic functionality and demonstrate
integration with C++, Python, and Bazel.

BUG=#2636

### CODE REVIEW

File: tensorflow/lite/micro/compression/metadata.fbs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace tflite.micro.compression;
+
+table Metadata {
+    // Compression data root, to be used in a tflite.Model.metadata field with
+    // the key "COMPRESSION_METADATA".
+
+    schema_version:int = 1;
```
Comment (suleshahid) [inline tensorflow/lite/micro/compression/metadata.fbs:21]: nit: uint

File: tensorflow/lite/micro/compression/metadata.fbs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    // compressed into indices into a table of values. The indices are unsigned
+    // integers, index_bitwidth-wide, in big-endian bit order, packed into the
+    // buffer identified by the corresponding tflite.Tensor's buffer field. The
+    // values are located in a newly-created buffer, encoded according to the
+    // tflite.Tensor.type. Tensors with multiple channels have distinct value
+    // tables for each channel, typically along their quantization axis,
+    // concatenated one after another. An element's index must be looked up in
+    // the value table corresponding to its channel.
+
+    tensor:int;            // index of the corresponding tflite.Tensor
```
Comment (suleshahid) [inline tensorflow/lite/micro/compression/metadata.fbs:51]: should this be uint as well? I believe tensor indices are never negative?

**Review by suleshahid — APPROVED** [review]

**Review by suleshahid — COMMENTED** [review]
oops meant to click comment not approve yet.

File: tensorflow/lite/micro/compression/metadata.fbs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+// limitations under the License.
+
+namespace tflite.micro.compression;
+
+table Metadata {
+    // Compression data root, to be used in a tflite.Model.metadata field with
+    // the key "COMPRESSION_METADATA".
+
+    schema_version:int = 1;
+        // ^ Incremented whenever there are backward-incompatible changes. Code
```
[the remark this answers is not in the document]
Comment (rkuester) [reply tensorflow/lite/micro/compression/metadata.fbs:22]: Will do.

File: tensorflow/lite/micro/compression/metadata.fbs (hunk 1 above)
Comment (rkuester) [reply tensorflow/lite/micro/compression/metadata.fbs]: Fixed in 13e34a0b.

File: tensorflow/lite/micro/compression/metadata.fbs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    lut_tensors:[LutTensor];
+        // ^ A list of tensors which are compressed using the
+        // (L)ook-(U)p-(T)able method. The indices of this vector are not
+        // significant.
+}
+
+table LutTensor {
+    // Look-Up-Table Tensor: a tensor representation where elements are
+    // compressed into indices into a table of values. The indices are unsigned
+    // integers, index_bitwidth-wide, in big-endian bit order, packed into the
```
[the remark this answers is not in the document]
Comment (rkuester) [reply tensorflow/lite/micro/compression/metadata.fbs]: Indeed flatbuffers are always little-endian *byte* order according to the flatbuffer spec. This comment was trying to explain the *bit* order of the index fields.

You're right that using big-endian to describe the bit order doesn't have a well-defined meaning in this context. I've taken a stab at improving this comment in db6179c9. What do you think?

File: tensorflow/lite/micro/compression/metadata.fbs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    // buffer identified by the corresponding tflite.Tensor's buffer field. The
+    // values are located in a newly-created buffer, encoded according to the
+    // tflite.Tensor.type. Tensors with multiple channels have distinct value
+    // tables for each channel, typically along their quantization axis,
+    // concatenated one after another. An element's index must be looked up in
+    // the value table corresponding to its channel.
+
+    tensor:int;            // index of the corresponding tflite.Tensor
+    value_buffer:uint;     // index of the buffer containing LUT values
```
[the remark this answers is not in the document]
Comment (rkuester) [reply tensorflow/lite/micro/compression/metadata.fbs:64]: That is correct. (Except that buffers are global, not per-subgraph.)

File: tensorflow/lite/micro/compression/metadata_test.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  metadata = schema.MetadataT()
+  metadata.subgraphs = [subgraph0]
+
+  # Build the flatbuffer itself using the flatbuffers runtime module.
+  builder = flatbuffers.Builder(32)
+  root = metadata.Pack(builder)
+  builder.Finish(root)
+  buffer: bytearray = builder.Output()
+
+  print(hexdump.hexdump(buffer, result='return'))
```
[the remark this answers is not in the document]
Comment (rkuester) [reply tensorflow/lite/micro/compression/metadata_test.py]: I did intentionally put those prints there for development and debugging. It's nice to have them in the test log if someone wants to see how the flatbuffer ended up looking in binary, and how much space it's consuming. The value of the output is not checked.

I've added some comments in the code to make that clear. Is to okay to leave them in under these circumstances?

File: tensorflow/lite/micro/compression/metadata_test.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  print(f"length: {len(buffer)}")
+
+  read_metadata = schema.MetadataT.InitFromPackedBuf(buffer, 0)
+  read_subgraph0 = read_metadata.subgraphs[0]
+
+  def attrs_equal(a, b):
+    return all(vars(a)[key] == vars(b)[key] for key in vars(a))
+
+  assert attrs_equal(read_subgraph0.lutTensors[0], lut_tensor0)
```
[the remark this answers is not in the document]
Comment (rkuester) [reply tensorflow/lite/micro/compression/metadata_test.py]: Fixed in 7ab3456b.

File: tensorflow/lite/micro/compression/metadata_test.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import sys
+import hexdump
+import flatbuffers
+
+# `.*_generated` is the name of the module created by the Bazel rule
+# `flatbuffer_py_library' based on the schema.
+from tensorflow.lite.micro.compression import metadata_py_generated as schema
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (rkuester) [reply tensorflow/lite/micro/compression/metadata_test.py]: Fixed in 7ab3456b. I used tf.test, as I've done in several other places. That looks to be another, simpler, name to the same thing. Is that just as well?

File: tensorflow/lite/micro/compression/metadata.fbs (hunk 4 above)
[the remark this answers is not in the document]
Comment (suleshahid) [reply tensorflow/lite/micro/compression/metadata.fbs]: Thanks for the more detailed comment. I think the padding part is slightly still confusing, but we should probably put a more detailed and more examples in a README, so this is fine for now.

**Comment by suleshahid** [comment]
Think that arm failure is aside of this change (https://github.com/tensorflow/tflite-micro/issues/2775#issuecomment-2481913933)

**Review by suleshahid — APPROVED** [review]

**Comment by suleshahid** [comment]
[@Mergifyio](https://github.com/Mergifyio) refresh

**Pull request merged**